docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+
+This proposal provides mechanisms for solving both the reliability and efficiency problems, by allowing pods to specify levels of resource guarantees. Containers will be able to *request* for a minimum resource guarantee. The *request* is different from the *limit* - containers will not be allowed to exceed resource limits.
+
+Additionally, we will support 3 different tiers of pods. Guaranteed pods (where requests and limits are equal) will receive the highest priority and will get all the resources they need. Burstable pods, which specify request < limit will be guaranteed their request and can “burst” and use more resources when available. Best-Effort pods, which don’t specify a request, can use resources only if not being used by other pods. In Borg, supporting best-effort pods increased utilization by about 20%, and we hope to see similar improvements in Kubernetes.
+
+## QoS Functionality
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity. If the container is successfully scheduled, the container is guaranteed the amount of resource requested. The container will not be allowed to exceed the specified limit. How the request and limit are enforced depends on whether the resource is [compressible or incompressible](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/resources.md).
+
+### Compressible Resource Guarantees
+- For now, we are only supporting CPU.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio.
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess resources when available.
+
+### Incompressible Resource Guarantees
+- For now, we are only supporting memory.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of memory they request, if they exceed their memory request, they could be killed (if some other job needs memory), but if they consume fewer resources than requested, they will not be killed.
+- Containers will be killed if they use more memory than their limit.

```

Comment:
It's true we'll have this problem once we kill the pod in an out-of-resource scenario. It's a special case of dealing with crash/failure loops, such as can be caused by an invalid Docker image #2529. We'll want controllers to back off in such cases.

To avoid this situation, we'll want a reactive vertical auto-sizer that intelligently grows resources when it detects repeated out-of-resource situations.